File: c#/LeadDatabase/Opportunities/DeleteOpportunityRoles.cs
/*
   DeleteOpportunityRoles.cs

   Marketo REST API Sample Code
   Copyright (C) 2016 Marketo, Inc.

   This software may be modified and distributed under the terms
   of the MIT license.  See the LICENSE file for details.
*/
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Net;
using System.IO;
using System.Collections;
using Newtonsoft.Json;
using System.Web;

namespace Samples
{

    class DeleteRoles
    {
        private String host = "CHANGE ME"; //host of your marketo instance, https://AAA-BBB-CCC.mktorest.com
        private String clientId = "CHANGE ME"; //clientId from admin > Launchpoint
        private String clientSecret = "CHANGE ME"; //clientSecret from admin > Launchpoint
        public String name;//name of custom object type
        public List<Dictionary<string, dynamic>> input;//list of records to delete
        public String deleteBy;//dedupeFields, or idFields, defaults to dedupeFields

        /*
        public static void Main(string[] args)
        {
            var delete = new DeleteRoles();
            delete.name = "pet";
            delete.input = new List<Dictionary<string, dynamic>>();
            var opp1 = new Dictionary<string, dynamic>();
            opp1.Add("externalopportunityid", "C# Test Opp");
            delete.input.Add(opp1);
            String result = delete.postData();
            Console.Write(result);
        }
        */

        public String postData()
        {
            //Assemble the URL
            String url = host + "/rest/v1/opportunities/roles/delete.json?access_token=" + getToken();
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
            request.Method = "POST";
            request.ContentType = "application/json";
            request.Accept = "application/json";
            StreamWriter wr = new StreamWriter(request.GetRequestStream());
            wr.Write(bodyBuilder());
            wr.Flush();
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Stream resStream = response.GetResponseStream();
            StreamReader reader = new StreamReader(resStream);
            return reader.ReadToEnd();
        }
        private String bodyBuilder()
        {
            var parent = new Dictionary<string, dynamic>();
            parent.Add("input", input);
            if (deleteBy != null)
            {
                parent.Add("deleteBy", deleteBy);
            }
            return JsonConvert.SerializeObject(parent);
        }

        private String getToken()
        {
            String url = host + "/identity/oauth/token?grant_type=client_credentials&client_id=" + clientId + "&client_secret=" + clientSecret;
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
            request.ContentType = "application/json";
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Stream resStream = response.GetResponseStream();
            StreamReader reader = new StreamReader(resStream);
            String json = reader.ReadToEnd();
            Dictionary<String, String> dict = JsonConvert.DeserializeObject<Dictionary<String, String>>(json);
            return dict["access_token"];
        }
    }
}

File: php/LeadDatabase/Opportunities/DeleteOpportunityRoles.php
<?php
/*
   DeleteOpportunityRoles.php

   Marketo REST API Sample Code
   Copyright (C) 2016 Marketo, Inc.

   This software may be modified and distributed under the terms
   of the MIT license.  See the LICENSE file for details.
*/
$delete = new DeleteOpportunityRoles();
$role1 = new stdClass();
$role1->externalopportunityid = "Opportunity 1";
$role1->role = "Captain";
$role1->leadId = 1;
$delete->roles = [$role1];
$delete->dedupeBy = "dedupeFields";
print_r($delete->postData());


class DeleteOpportunityRoles{
	private $host = "CHANGE ME";
	private $clientId = "CHANGE ME";
	private $clientSecret = "CHANGE ME";
	public $roles;//array of objects with externalopportunityid, leadid and role
	public $idfields; //array of marketo IDs
	public $dedupeBy; //dedupe field, dedupeFields or idField
	
	private function bodyBuilder(){
		$requestBody = new stdClass();
		//set dedupeby parameter in json body
		$requestBody->dedupeBy = $this->dedupeBy;
		$requestBody->input = array();
		$i = 0;
		//if dedupeby is dedupefields, use externalopportunityid
		if ($this->dedupeBy === "dedupeFields"){
			$requestBody->input = $this->roles;
		}//else use marketoGUID
		else if ($this->dedupeBy === "idField"){
			foreach($this->input as $id){
				$obj = new stdClass();
				$obj->marketoGUID = $id;
				$requestBody->input[$i] = $obj;
				$i++;
			}
		}
		$json = json_encode($requestBody);
		return $json;
	}
	
	public function postData(){
		$url = $this->host . "/rest/v1/opportunities/delete.json?access_token=" . $this->getToken();
		$ch = curl_init($url);
		$requestBody = $this->bodyBuilder();
		print_r($requestBody);
		curl_setopt($ch,  CURLOPT_RETURNTRANSFER, 1);
		curl_setopt($ch, CURLOPT_HTTPHEADER, array('accept: application/json','Content-Type: application/json'));
		curl_setopt($ch, CURLOPT_POST, 1);
		curl_setopt($ch, CURLOPT_POSTFIELDS, $requestBody);
		curl_getinfo($ch);
		$response = curl_exec($ch);
		return $response;
	}
	
	private function getToken(){
		$ch = curl_init($this->host . "/identity/oauth/token?grant_type=client_credentials&client_id=" . $this->clientId . "&client_secret=" . $this->clientSecret);
		curl_setopt($ch,  CURLOPT_RETURNTRANSFER, 1);
		curl_setopt($ch, CURLOPT_HTTPHEADER, array('accept: application/json',));
		$response = json_decode(curl_exec($ch));
		curl_close($ch);
		$token = $response->access_token;
		return $token;
	}	
}